### ISSUE TITLE
Tutorial: Memory Pool in C

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10488]
Tutorial: Memory Pool in C++<br>
<br>
https://ift.tt/B3loigG<br>
<br>
Mario Konrad<br>
<br>

<div><div><div><div>
<div>
<p>Tutorial: Memory Pool in C++</p>
<p>
©
2010
/ Mario Konrad
</p>
<h2>Introduction</h2>
<p>This tutorial shows how to use memory pools to increase memory allocation/deallocation and preventing fragmentation. Useful in high performance systems or systems with limited resources, e.g. embedded and/or realtime systems.</p>
<p>It is implemented in C++ using templates.</p>
<p>Please note: this is not a beginners tutorial. Knowledge of the programming language and development environment are assumed.</p>
<p>A memory pool is, in general, a preallocated chunk of memory to be used by an application. The preallocation prevents the application to ask the operating system for memory (through <code>new</code>/<code>delete</code> or <code>malloc</code>/<code>free</code>). This way, preallocated memory saves system calls and defies the non-deterministic nature of memory allocation.</p>
<p>To prevent fragmentation, the easy solution is to have every object within the memory pool the same size. This sounds like an unnatural requirement but in practice (embedded realtime systems) the communication within the application is often done using messages of a defined size.</p>
<p>In practice it is often useful to define multiple memory pools for different kind of messages or for different communication channels (e.g. synchronization points like message queues).</p>
<p>The memory pool shown in this tutorial does the exactly the things stated above.</p>
<p><strong>Note:</strong> to make the tutorial focus on the imporant parts, this implementation is not thread-safe.</p>
<h2>In Practice</h2>
<p>Considering the following code, it is a simple application of dynamic memory usage, creating object from a class <code>A</code>:</p>
<p>The usage of a memory pool should be transparent for the rest of the application. Let’s have a look at the class <code>A</code>. For easier inspection, the code is written inline.</p>
<p>This code already works but does not use a memory pool, it uses the standard <code>new</code>/<code>delete</code> mechanism for memory management.</p>
<p>The following code shows the extension to be made to use the memory pool for objects of the class <code>A</code>. All other classes are still handled by standard memory management.</p>
<p>The memory pool is static for the class A, so all objects use the memory pool. Other classes are not affected. From this point it is not possible to have some objects of this class in a memory pool and some not.</p>
<h2>Memory Pool</h2>
<p>The implementation of the memory pool is rather simple:</p>
<p>The important and interesting part is done in the <code>Strategy</code>. The design is intentional, to make the behaviour of the memory pool configurable.</p>
<p><strong>Note:</strong> the memory pool (or the strategy respectively) is initialized only once. Since the pool is a static member of the class <code>A</code>, the constructor of the memory pool is called only once.</p>
<p>The next chapter gets into the different strategies.</p>
<h2>Strategies</h2>
<p>This chapter shows different strategies to configure the memoy pool. This way it is very easy to extend the memory pool with different strategies.</p>
<h3>malloc/free</h3>
<p>This strategy is simply to use the default memory management, using <code>malloc</code> and <code>free</code>.</p>
<p>There is nothing to initialize (may be different in certain embedded operating systems). The allocation and deallocation is done by calling <code>malloc</code> and <code>free</code>.</p>
<p>The template parameter simply takes the type of data to be managed.</p>
<p>A thing to note: using this strategy the memory pool is not really a pool anymore. Also the memory is limited by what has the operating system to offer.</p>
<h2>Simple Array</h2>
<p>A simple array to handle memory is not a very good idea in practice, but it helps to show different types of strategies and for performance comparsions later on.</p>
<p>The template parameters are <code>T</code> the type of data to be managed and N the number of memory chunks to be reserved for the memory pool to be used. It is not possible to use more objects of the class <code>T</code> than <code>N</code>.</p>
<p>The allocation of the memory is done in a linear search through the array to find an unoccupied place. If the memory pool is very small, this might even be reasonable. However not recommended for larger memory pools.</p>
<p>The deallocation is the same: search for the place to be freed and do it.</p>
<h3>Heap</h3>
<p>This strategy uses a heap (see <a href="http://en.wikipedia.org/wiki/Heap_%28data_structure%29">Heap @ wikipedia</a>) to handle the memory pool. Operations (allocation and deallocation) are of logarithmic complexity. It is not part of this tutorial to show how a heap works.</p>
<div><pre><code><a title="1"><span>template</span> &lt;<span>typename</span> T, <span>unsigned</span> <span>int</span> N&gt; <span>class</span> mempool_heap</a>
<a title="2">{</a>
<a title="3"> <span>private</span>:</a>
<a title="4"> <span>typedef</span> <span>unsigned</span> <span>int</span> <span>size_type</span>; <span>// convinience</span></a>
<a title="5"> <span>private</span>:</a>
<a title="6"> <span>size_type</span> available; <span>// number of memory chunks available</span></a>
<a title="7"> T * a[N+<span>1</span>]; <span>// book keeping, first place not used</span></a>
<a title="8"> <span>uint8_t</span> buf[<span>sizeof</span>(T)*N]; <span>// the memory itself, here will the objects be stored</span></a>
<a title="9"> <span>private</span>:</a>
<a title="10"> <span>inline</span> <span>void</span> swap(<span>size_type</span> i, <span>size_type</span> j)</a>
<a title="11"> {</a>
<a title="12"> T * t = a[i];</a>
<a title="13"> a[i] = a[j];</a>
<a title="14"> a[j] = t;</a>
<a title="15"> }</a>
<a title="16"></a>
<a title="17"> <span>void</span> up()</a>
<a title="18"> {</a>
<a title="19"> <span>for</span> (<span>size_type</span> n = available; n &gt; <span>1</span>; ) {</a>
<a title="20"> <span>size_type</span> i = n / <span>2</span>;</a>
<a title="21"> <span>if</span> (!(a[i] &amp;&amp; a[n]))</a>
<a title="22"> swap(i, n);</a>
<a title="23"> n = i;</a>
<a title="24"> }</a>
<a title="25"> }</a>
<a title="26"></a>
<a title="27"> <span>void</span> down()</a>
<a title="28"> {</a>
<a title="29"> <span>size_type</span> i = <span>1</span>;</a>
<a title="30"> <span>size_type</span> c = <span>2</span> * i;</a>
<a title="31"> <span>while</span> (c &lt;= available) {</a>
<a title="32"> <span>if</span> (c+<span>1</span> &lt;= available)</a>
<a title="33"> <span>if</span> (!(a[c] &amp;&amp; a[c+<span>1</span>]))</a>
<a title="34"> ++c;</a>
<a title="35"> <span>if</span> (!(a[i] &amp;&amp; a[c]))</a>
<a title="36"> swap(i, c);</a>
<a title="37"> i = c;</a>
<a title="38"> c = <span>2</span> * i;</a>
<a title="39"> }</a>
<a title="40"> }</a>
<a title="41"> <span>public</span>:</a>
<a title="42"> <span>void</span> init()</a>
<a title="43"> {</a>
<a title="44"> <span>// number of available memory chunks is the size of the memory pool, naturally</span></a>
<a title="45"> available = N;</a>
<a title="46"></a>
<a title="47"> <span>// first node is not used, convinience</span></a>
<a title="48"> a[<span>0</span>] = NULL;</a>
<a title="49"></a>
<a title="50"> <span>// all other memory chunks are free, the pointers are initialized</span></a>
<a title="51"> <span>for</span> (<span>size_type</span> i = <span>1</span>; i &lt;= N; ++i) {</a>
<a title="52"> a[i] = <span>reinterpret_cast</span>&lt;T *&gt;(&amp;buf[<span>sizeof</span>(T)*(i-<span>1</span>)]);</a>
<a title="53"> }</a>
<a title="54"> }</a>
<a title="55"></a>
<a title="56"> <span>void</span> * allocate()</a>
<a title="57"> {</a>
<a title="58"> <span>// allocation not possible if the memory pool has no more space</span></a>
<a title="59"> <span>if</span> (available &lt;= <span>0</span>) <span>throw</span> <span>std::</span>bad_alloc();</a>
<a title="60"></a>
<a title="61"> <span>// the first memory chunk is always on index 1, this is guaranteed</span></a>
<a title="62"> <span>// by the heap</span></a>
<a title="63"> T * t = a[<span>1</span>];</a>
<a title="64"></a>
<a title="65"> <span>// move the last free node to front</span></a>
<a title="66"> a[<span>1</span>] = a[available];</a>
<a title="67"></a>
<a title="68"> <span>// mark the used chunk</span></a>
<a title="69"> a[available] = NULL;</a>
<a title="70"></a>
<a title="71"> <span>// number of free chunks decreased</span></a>
<a title="72"> --available;</a>
<a title="73"></a>
<a title="74"> <span>// book keeping for the heap</span></a>
<a title="75"> down();</a>
<a title="76"></a>
<a title="77"> <span>// return a pointer to the allocated space</span></a>
<a title="78"> <span>return</span> t;</a>
<a title="79"> }</a>
<a title="80"></a>
<a title="81"> <span>void</span> deallocate(<span>void</span> * p)</a>
<a title="82"> {</a>
<a title="83"> <span>// invalid pointers are ignored</span></a>
<a title="84"> <span>if</span> (!p || available &gt;= N) <span>return</span>;</a>
<a title="85"></a>
<a title="86"> <span>// memory is freed, more space within the pool</span></a>
<a title="87"> ++available;</a>
<a title="88"></a>
<a title="89"> <span>// the freed node is, well, free. status and pointer are restored.</span></a>
<a title="90"> a[available] = <span>reinterpret_cast</span>&lt;T *&gt;(p);</a>
<a title="91"></a>
<a title="92"> <span>// book keeping for the heap</span></a>
<a title="93"> up();</a>
<a title="94"> }</a>
<a title="95">};</a></code></pre></div>
<p>The template parameters are the same as in chapter [[#Simple Array|Simple Array]]: data type and size of the memory pool.</p>
<p>Since the operations using the heap are of logarithmic nature, this strategy is very useful for large memory pools.</p>
<h3>Standard Template Library Heap</h3>
<p>The C++ Standard Library already provides functions to use heaps. This strategy is also a heap, using the functions provided by the standard library (<code>make_heap</code>, <code>pop_heap</code>, <code>push_heap</code>).</p>
<p>The implementation is very similar to the heap above. The disadvantage is the dependency to the <code>algorithm</code> header of the standard library.</p>
<div><pre><code><a title="1"><span>template</span> &lt;<span>typename</span> T, <span>unsigned</span> <span>int</span> N&gt; <span>class</span> mempool_std_heap</a>
<a title="2">{</a>
<a title="3"> <span>private</span>:</a>
<a title="4"> <span>// states of the heap nodes</span></a>
<a title="5"> <span>enum</span> State { FREE = <span>1</span>, TAKEN = <span>0</span> };</a>
<a title="6"></a>
<a title="7"> <span>struct</span> Entry {</a>
<a title="8"> State state; <span>// the state of the memory chunk</span></a>
<a title="9"> T * p; <span>// pointer to the memory chunk</span></a>
<a title="10"></a>
<a title="11"> <span>// comparsion operator, needed for heap book keeping</span></a>
<a title="12"> <span>bool</span> <span>operator</span>&lt;(<span>const</span> Entry &amp; other) <span>const</span></a>
<a title="13"> { <span>return</span> state &lt; other.state; }</a>
<a title="14"> };</a>
<a title="15"></a>
<a title="16"> <span>typedef</span> <span>unsigned</span> <span>int</span> <span>size_type</span>; <span>// convenience</span></a>
<a title="17"> <span>private</span>:</a>
<a title="18"> <span>size_type</span> available; <span>// number of memory chunks available</span></a>
<a title="19"> Entry a[N]; <span>// book keeping</span></a>
<a title="20"> <span>uint8_t</span> buf[<span>sizeof</span>(T)*N]; <span>// the actual memory, here will the objects be stored</span></a>
<a title="21"> <span>public</span>:</a>
<a title="22"> <span>void</span> init()</a>
<a title="23"> {</a>
<a title="24"> <span>// number of available memory chunks is the size of the memory pool</span></a>
<a title="25"> available = N;</a>
<a title="26"></a>
<a title="27"> <span>// all memory chunks are free, pointers are initialized</span></a>
<a title="28"> <span>for</span> (<span>size_type</span> i = <span>0</span>; i &lt; N; ++i) {</a>
<a title="29"> a[i].state = FREE;</a>
<a title="30"> a[i].p = <span>reinterpret_cast</span>&lt;T *&gt;(&amp;buf[<span>sizeof</span>(T)*i]);</a>
<a title="31"> }</a>
<a title="32"></a>
<a title="33"> <span>// make heap of book keeping array</span></a>
<a title="34"> <span>std::</span>make_heap(a, a+N);</a>
<a title="35"> }</a>
<a title="36"></a>
<a title="37"> <span>void</span> * allocate()</a>
<a title="38"> {</a>
<a title="39"> <span>// allocation not possible if memory pool has no more space</span></a>
<a title="40"> <span>if</span> (available &lt;= <span>0</span> || available &gt; N) <span>throw</span> <span>std::</span>bad_alloc();</a>
<a title="41"></a>
<a title="42"> <span>// the first memory chunk is always on index 0</span></a>
<a title="43"> Entry e = a[<span>0</span>];</a>
<a title="44"></a>
<a title="45"> <span>// remove first entry from heap</span></a>
<a title="46"> <span>std::</span>pop_heap(a, a+N);</a>
<a title="47"></a>
<a title="48"> <span>// one memory chunk allocated, no more available</span></a>
<a title="49"> --available;</a>
<a title="50"></a>
<a title="51"> <span>// mark the removed chunk</span></a>
<a title="52"> a[available].state = TAKEN;</a>
<a title="53"> a[available].p = NULL;</a>
<a title="54"></a>
<a title="55"> <span>// return pointer to the allocated memory</span></a>
<a title="56"> <span>return</span> e.p;</a>
<a title="57"> }</a>
<a title="58"></a>
<a title="59"> <span>void</span> deallocate(<span>void</span> * ptr)</a>
<a title="60"> {</a>
<a title="61"> <span>// invalid pointers are ignored</span></a>
<a title="62"> <span>if</span> (!ptr || available &gt;= N) <span>return</span>;</a>
<a title="63"></a>
<a title="64"> <span>// mark freed memory as such</span></a>
<a title="65"> a[available].state = FREE;</a>
<a title="66"> a[available].p = <span>reinterpret_cast</span>&lt;T *&gt;(ptr);</a>
<a title="67"></a>
<a title="68"> <span>// freed memory chunk, one more available</span></a>
<a title="69"> ++available;</a>
<a title="70"></a>
<a title="71"> <span>// heap book keeping</span></a>
<a title="72"> <span>std::</span>push_heap(a, a+N);</a>
<a title="73"> }</a>
<a title="74">};</a></code></pre></div>
<h2>Performance</h2>
<p>This chapter compares the memory pool strategies shown above.</p>
<p><strong>TODO</strong></p>
<h2>Download</h2>
<p>All files listed for download are free to modify, distribute or to use them in any kind you like.</p>
<p>Use them on your own risk.</p>
<p>One package:</p>
</div>
</div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via mario::konrad - Contents <br>
<br>
December 28, 2023 at 03:50PM
